src/sources/exec.rs

```diff
@@ -0,0 +1,830 @@
+    // this seems to be the best approximation of exec duration
+    let start = Instant::now();
+
+    let mut child = command.spawn()?;
+
+    let stdout = child
+        .stdout
+        .take()
+        .ok_or_else(|| Error::new(ErrorKind::Other, "Unable to take stdout of spawned process"))?;
+
+    // Create stdout async reader
+    let stdout_reader = BufReader::new(stdout);
+
+    // Set up communication channels
+    let (sender, mut receiver) = channel(1024);
+
+    let pid = child.id();
+
+    // Optionally include stderr
+    if config.include_stderr {
+        let stderr = child.stderr.take().ok_or_else(|| {
+            Error::new(ErrorKind::Other, "Unable to take stderr of spawned process")
+        })?;
+
+        // Create stderr async reader
+        let stderr_reader = BufReader::new(stderr);
+
+        spawn_reader_thread(
+            stderr_reader,
+            shutdown.clone(),
+            config.event_per_line,
+            config.maximum_buffer_size,
+            STDERR,
+            sender.clone(),
+        );
+    }
+
+    spawn_reader_thread(
+        stdout_reader,
+        shutdown.clone(),
+        config.event_per_line,
+        config.maximum_buffer_size,
+        STDOUT,
+        sender,
+    );
+
+    while let Some((line, stream)) = receiver.recv().await {
+        let event = create_event(
+            &config,
+            &hostname,
+            line,
+            &Some(stream.to_string()),
+            pid,
+            None,
+            &None,
+        );
+
+        let _ = out
+            .send(event)
+            .await
+            .map_err(|_: crate::pipeline::ClosedError| {
+                error!(message = "Failed to forward events; downstream is closed.");
+            });
+    }
+
+    let elapsed = start.elapsed();
+
+    info!("Finished command run.");
+    let _ = out.flush().await;
+
+    match child.try_wait() {
+        Ok(Some(exit_status)) => {
+            handle_exit_status(&config, hostname, pid, Some(exit_status), elapsed, out).await;
+            Ok(Some(exit_status))
+        }
+        Ok(None) => {
+            handle_exit_status(&config, hostname, pid, None, elapsed, out).await;
+            Ok(None)
+        }
+        Err(error) => {
+            error!(message = "Unable to obtain exit status.", %error);
+
+            handle_exit_status(&config, hostname, pid, None, elapsed, out).await;
+            Ok(None)
+        }
+    }
+}
+
+async fn handle_exit_status(
+    config: &ExecConfig,
+    hostname: Option<String>,
+    pid: Option<u32>,
+    exit_status: Option<ExitStatus>,
+    exec_duration: Duration,
+    mut out: Pipeline,
+) {
+    let exit_status = match exit_status {
+        Some(exit_status) => exit_status.code(),
+        None => None,
+    };
+
+    emit!(ExecCommandExecuted {
+        command: config.command_line().as_str(),
+        exit_status,
+        exec_duration,
+    });
+
+    let event = create_event(
```

Comment:
I'm still on the fence about sending this final event since it doesn't actually correspond to a line from the output. Do you think it is necessary?

Reply:
My thought is that it is nice to sweep up the `exit_status` and `exec_duration_millis` and provide it in an event if we have it. It gives the downstream transformers something to act on if they want it. Just having the internal event (`ExecCommandExecuted`) might be ok but I was not sure. Conversely it might be a pain for downstream consumers to strip out if they are not interested. To be honest, I am really not sure. Happy to remove it and or keep it based on your final decision.

Reply:
Removed